### PR TITLE
feat: implement json for proof

### PR DESCRIPTION
**Pull request by dongchangYoo**
In this PR, `RapidJsonValueConverters` for `Proof` implemented in `zk/plonk/halo2/proof.h` has been implemented.
From now on, the `Proof` can be written to a JSON file and read from the JSON file.

### CODE REVIEW

File: tachyon/zk/plonk/halo2/proof_unittest.cc [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+class ProofTest : public testing::Test {
+ public:
+  constexpr static size_t kMaxDegree = (size_t{1} << 5) - 1;
+  constexpr static size_t kMaxExtendedDegree = (size_t{1} << 7) - 1;
+
+  using PCS = SHPlonkExtension<math::bn254::BN254Curve, kMaxDegree,
+                               kMaxExtendedDegree, math::bn254::G1AffinePoint>;
+  using F = PCS::Field;
+  using Commitment = PCS::Commitment;
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/proof_unittest.cc:26]: Please just use `bn254::Fr`, and `bn254::G1AffinePoint`

Comment (chokobole) [reply tachyon/zk/plonk/halo2/proof_unittest.cc:26]: You forgot to do `math::bn254::G1Curve::Init()`

File: tachyon/zk/plonk/halo2/proof_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+ protected:
+  size_t num_circuits_ = 2;
+  size_t num_elements_ = 3;
+};
+
+}  // namespace
+
+TEST_F(ProofTest, JsonValueConverter) {
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/proof_unittest.cc:52]: Please just test it without fixture.

**Review by chokobole — CHANGES_REQUESTED** [review]

File: tachyon/zk/plonk/halo2/proof_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+#include "tachyon/zk/base/commitments/shplonk_extension.h"
+
+namespace tachyon::zk::halo2 {
+namespace {
+
+using F = math::bn254::Fr;
+using Commitment = math::bn254::G1AffinePoint;
+
+template <typename T>
+static std::vector<std::vector<T>> CreateRandomElementsVec(size_t rows,
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/proof_unittest.cc:22]:
```suggestion
std::vector<std::vector<T>> CreateRandomPrimeFieldsVec(size_t rows,
```

File: tachyon/zk/plonk/halo2/proof_unittest.cc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  EXPECT_EQ(proof.lookup_product_evals_vec,
+            expected_proof.lookup_product_evals_vec);
+  EXPECT_EQ(proof.lookup_product_next_evals_vec,
+            expected_proof.lookup_product_next_evals_vec);
+  EXPECT_EQ(proof.lookup_permuted_input_evals_vec,
+            expected_proof.lookup_permuted_input_evals_vec);
+  EXPECT_EQ(proof.lookup_permuted_input_inv_evals_vec,
+            expected_proof.lookup_permuted_input_inv_evals_vec);
+  EXPECT_EQ(proof.lookup_permuted_table_evals_vec,
+            expected_proof.lookup_permuted_table_evals_vec);
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/proof_unittest.cc:134]: How about implementing `operator==()` for `Proof`?

File: tachyon/zk/plonk/halo2/proof.h [hunk 5]
```diff
@@ -49,6 +49,42 @@ struct Proof {
...
+           lookup_product_next_evals_vec ==
+               other.lookup_product_next_evals_vec &&
+           lookup_permuted_input_evals_vec ==
+               other.lookup_permuted_input_evals_vec &&
+           lookup_permuted_input_inv_evals_vec ==
+               other.lookup_permuted_input_inv_evals_vec &&
+           lookup_permuted_table_evals_vec ==
+               other.lookup_permuted_table_evals_vec;
+  }
+
```
Comment (chokobole) [inline tachyon/zk/plonk/halo2/proof.h:90]: Please add `operator!=()` as well!

**Review by chokobole — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**